trajopt/src/trajectory_costs.cpp

```diff
@@ -18,30 +18,227 @@ namespace trajopt
+      expr_vec_.push_back(expr_neg);
+    }
+  }
+}
+
+double JointPosIneqCost::value(const DblVec& xvec)
+{
+  // Convert vector from optimization to trajectory
+  Eigen::MatrixXd traj = getTraj(xvec, vars_);
+  // Takes diff b/n the subsequent rows to get velocity
+  Eigen::MatrixXd pos = traj.block(first_step_, 0, last_step_ - first_step_ + 1, traj.cols());
+  // Subtract targets to center about 0 and then subtract from tolerance
+  Eigen::MatrixXd diff0 = (pos.rowwise() - targets_.transpose());
+  Eigen::MatrixXd diff1 = (diff0.rowwise() - upper_tols_.transpose()) * coeffs_.asDiagonal();
+  Eigen::MatrixXd diff2 = ((diff0 * -1).rowwise() + lower_tols_.transpose()) * coeffs_.asDiagonal();
+  // Applies hinge, multiplies it by a diagonal matrix of coefficients, sums each corresponding value, and converts to
+  // vector
+  return diff1.cwiseMax(0).sum() + diff2.cwiseMax(0).sum();
+}
+
+sco::ConvexObjectivePtr JointPosIneqCost::convex(const DblVec& /*x*/, sco::Model* model)
+{
+  sco::ConvexObjectivePtr out(new sco::ConvexObjective(model));
+  // Add hinge cost. Set the coefficient to 1 here since we include it in the AffExpr already
+  // This is necessary since we want a seperate coefficient per joint
+  for (sco::AffExpr expr : expr_vec_)
+  {
+    out->addHinge(expr, 1);
+  }
+  return out;
+}
+
+JointPosEqConstraint::JointPosEqConstraint(const VarArray& vars,
+                                           const Eigen::VectorXd& coeffs,
+                                           const Eigen::VectorXd& targets,
+                                           int& first_step,
+                                           int& last_step)
+  : EqConstraint("JointPosEq")
+  , vars_(vars)
+  , coeffs_(coeffs)
+  , targets_(targets)
+  , first_step_(first_step)
+  , last_step_(last_step)
+{
+  for (int i = first_step_; i <= last_step_; ++i)
+  {
+    for (int j = 0; j < vars.cols(); ++j)
+    {
+      // pos = x1 - targ
+      sco::AffExpr pos;
+      sco::exprInc(pos, sco::exprMult(vars(i, j), 1));
+      sco::exprDec(pos, targets_[j]);
+      // expr_ = coeff * vel - Not squared b/c QuadExpr cnt not yet supported (TODO)
+      expr_vec_.push_back(sco::exprMult(pos, coeffs_[j]));
+    }
+  }
+}
+
+DblVec JointPosEqConstraint::value(const DblVec& xvec)
+{
+  // Convert vector from optimization to trajectory
+  Eigen::MatrixXd traj = getTraj(xvec, vars_);
+  // Takes diff b/n the subsequent rows and subtract each row by the targets_ vector
+  Eigen::MatrixXd diff =
+      (traj.block(first_step_, 0, last_step_ - first_step_ + 1, traj.cols())).rowwise() - targets_.transpose();
+  // Squares it, multiplies it by a diagonal matrix of coefficients, and converts to vector
+  return util::toDblVec((diff.array().square()).matrix() * coeffs_.asDiagonal());
+}
+sco::ConvexConstraintsPtr JointPosEqConstraint::convex(const DblVec& /*x*/, sco::Model* model)
+{
+  sco::ConvexConstraintsPtr out(new sco::ConvexConstraints(model));
+  for (sco::AffExpr expr : expr_vec_)
```

Comment:
Should we be making copies?
```suggestion
  for (sco::AffExpr& expr : expr_vec_)
```